### ISSUE TITLE
what is the difference between the model and para_model in prepare_model?

### ISSUE THREAD

**Issue by seyeeet** [Sha-Lab/FEAT#42]
I see in the prepare_model we have:
```
def prepare_model(args):
    model = eval(args.model_class)(args)

    # load pre-trained model (no FC weights)
    if args.init_weights is not None:
        model_dict = model.state_dict()        
        pretrained_dict = torch.load(args.init_weights)['params']
        if args.backbone_class == 'ConvNet':
            pretrained_dict = {'encoder.'+k: v for k, v in pretrained_dict.items()}
        pretrained_dict = {k: v for k, v in pretrained_dict.items() if k in model_dict}
        print(pretrained_dict.keys())
        model_dict.update(pretrained_dict)
        model.load_state_dict(model_dict)

    if torch.cuda.is_available():
        torch.backends.cudnn.benchmark = True
        
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = model.to(device)
    if args.multi_gpu:
        model.encoder = nn.DataParallel(model.encoder, dim=0)
        para_model = model.to(device)
    else:
        para_model = model.to(device)

    return model, para_model
```
later we have 
```
def train(self):
        args = self.args
        self.model.train()
        if self.args.fix_BN:
            self.model.encoder.eval()
        
        # start FSL training
        label, label_aux = self.prepare_label()
        for epoch in range(1, args.max_epoch + 1):
            self.train_epoch += 1
            self.model.train()
            if self.args.fix_BN:
                self.model.encoder.eval()
            
            tl1 = Averager()
            tl2 = Averager()
            ta = Averager()

            start_tm = time.time()
            for batch in self.train_loader:
                self.train_step += 1

                if torch.cuda.is_available():
                    data, gt_label = [_.cuda() for _ in batch]
                else:
                    data, gt_label = batch[0], batch[1]
               
                data_tm = time.time()
                self.dt.add(data_tm - start_tm)

                # get saved centers
                logits, reg_logits = self.para_model(data)
                if reg_logits is not None:
                    loss = F.cross_entropy(logits, label)
                    total_loss = loss + args.balance * F.cross_entropy(reg_logits, label_aux)
                else:
                    loss = F.cross_entropy(logits, label)
                    total_loss = F.cross_entropy(logits, label)
```
what is the difference between `self.para_model` and `self.model`, and why we have  `self.model.train()` at the beginning , but in the   `prepare_model(args)` we have `model = eval(args.model_class)(args)`, why we use `eval` here and then `.train()`. can you please explain it?
Thanks a lot

**Comment by Han-Jia** [comment]
the para_model is used as a parallel interface of the model, which is set the same as the model in our implementation. 

eval in "eval(args.model_class)(args) " is to call the init function of the model class, if we set args.model_class = ProtoNet, it is same as "Protonet(args)". The eval in "model.eval()" is to set the model to the evaluation mode.

**Comment by z1358** [comment]
So when using a single GPU, `logits, reg_logits = self.para_model(data)` is the same as `logits, reg_logits = self.model(data)`?
Thanks a lot.
